src/node_url.cc

```diff
@@ -75,6 +75,73 @@ void BindingData::Deserialize(v8::Local<v8::Context> context,
   CHECK_NOT_NULL(binding);
 }
 
+std::string EncodePathChars(const std::string& input_str, bool windows) {
+  std::ostringstream encoded;
+  encoded << "file://";
+  for (char i : input_str) {
+    switch (i) {
+#define URL_ENCODE(char, code)                                                 \
+  case char:                                                                   \
+    encoded << code;                                                           \
+    break;
+
+      URL_ENCODE('%', "%25");
+      URL_ENCODE('\t', "%09");
+      URL_ENCODE('\n', "%0A");
+      URL_ENCODE('\r', "%0D");
+      URL_ENCODE(' ', "%20");
+      URL_ENCODE('"', "%22");
+      URL_ENCODE('#', "%23");
+      URL_ENCODE('?', "%3F");
+      URL_ENCODE('[', "%5B");
+      URL_ENCODE(']', "%5D");
+      URL_ENCODE('^', "%5E");
+      URL_ENCODE('|', "%7C");
+      URL_ENCODE('~', "%7E");
+#undef URL_ENCODE
+
+      case '\\':
+        if (!windows) {
+          encoded << "%5C";
+          break;
+        }
+      // fallthrough
+      default:
+        encoded << i;  // Append the character as is
+        break;
+    }
+  }
+
+  return encoded.str();
+}
+
+void BindingData::PathToFileURL(const FunctionCallbackInfo<Value>& args) {
+  CHECK_GE(args.Length(), 2);  // input
+  CHECK(args[0]->IsString());
+  CHECK(args[1]->IsBoolean());
+
+  Realm* realm = Realm::GetCurrent(args);
+  BindingData* binding_data = realm->GetBindingData<BindingData>();
+  Isolate* isolate = realm->isolate();
+  auto windows = args[1]->IsTrue();
+
+  Utf8Value input(isolate, args[0]);
+  auto input_str = input.ToString();
```

Comment:
We should avoid extra allocation:

```suggestion
  auto input_str = input.ToStringView();
```